File: solutions/java/meetup/3/src/main/java/Meetup.java
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.temporal.TemporalAdjusters;

record Meetup(int month, int year) {
    LocalDate day(DayOfWeek dayOfWeek, MeetupSchedule schedule) {
        var base = switch(schedule) {
            case TEENTH -> LocalDate.of(year, month, 13);
            default -> LocalDate.of(year, month, 1);
        };
        var modifier = switch(schedule) {
            case FIRST -> TemporalAdjusters.dayOfWeekInMonth(1, dayOfWeek);
            case SECOND -> TemporalAdjusters.dayOfWeekInMonth(2, dayOfWeek);
            case THIRD -> TemporalAdjusters.dayOfWeekInMonth(3, dayOfWeek);
            case FOURTH -> TemporalAdjusters.dayOfWeekInMonth(4, dayOfWeek);
            case LAST -> TemporalAdjusters.dayOfWeekInMonth(-1, dayOfWeek);
            case TEENTH -> TemporalAdjusters.nextOrSame(dayOfWeek);
        };
        return base.with(modifier);
    }
}

File: solutions/haskell/meetup/3/src/Meetup.hs
module Meetup (Weekday(..), Schedule(..), meetupDay) where

import Data.Time.Calendar (Day, firstDayOfWeekOnAfter, fromGregorian, gregorianMonthLength)

data Weekday = Monday | Tuesday | Wednesday | Thursday | Friday | Saturday | Sunday deriving Enum

data Schedule = First | Second | Third | Fourth | Last | Teenth deriving Enum

meetupDay :: Schedule -> Weekday -> Integer -> Int -> Day
meetupDay sch wd y m = firstDayOfWeekOnAfter (toEnum $ fromEnum wd + 1) (fromGregorian y m d)
  where
    d = case sch of
      Last -> gregorianMonthLength y m - 6
      Teenth -> 13
      s -> 1 + 7 * fromEnum s